pkg/client/cache/reflector.go

```diff
@@ -160,16 +180,47 @@ var (
 // required, and a cleanup function.
 func (r *Reflector) resyncChan() (<-chan time.Time, func() bool) {
 	if r.resyncPeriod == 0 {
+		r.nextResync = time.Time{}
 		return neverExitWatch, func() bool { return false }
 	}
 	// The cleanup function is required: imagine the scenario where watches
 	// always fail so we end up listing frequently. Then, if we don't
 	// manually stop the timer, we could end up with many timers active
 	// concurrently.
+	r.nextResync = now().Add(r.resyncPeriod)
 	t := time.NewTimer(r.resyncPeriod)
 	return t.C, t.Stop
 }
 
+// We want to avoid situations when periodic resyncing is breaking the TCP
+// connection.
+// If response`s body is not read to completion before calling body.Close(),
+// that TCP connection will not be reused in the future - see #15664 issue
+// for more details.
```

Comment:
Anything http Hijacked can't be returned to the pool and is closed at the
end

On Oct 27, 2015, at 12:29 AM, Daniel Smith <notifications@github.com> wrote:

In pkg/client/cache/reflector.go
<https://github.com/kubernetes/kubernetes/pull/16266#discussion_r43083521>:

>  	t := time.NewTimer(r.resyncPeriod)
>  	return t.C, t.Stop
>  }
>
> +// We want to avoid situations when periodic resyncing is breaking the TCP
> +// connection.
> +// If response`s body is not read to completion before calling body.Close(),
> +// that TCP connection will not be reused in the future - see #15664 issue
> +// for more details.

Did you consider using http.Hijacker, getting the raw net.Conn, closing the
write half, and waiting for the server to then close the other half? (not
sure if it will work but would perhaps be simpler than this)

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16266/files#r43083521>.